docs/concepts/overview/extending.md

```diff
@@ -0,0 +1,217 @@
+
+**Built-in Policy APIs**, such as [ResourceQuota](/docs/concepts/policy/resource-quotas/), [PodSecurityPolicies](/docs/concepts/policy/pod-security-policy/), [NetworkPolicy](/docs/concepts/services-networking/network-policies/) and Role-based Access Control ([RBAC](/docs/admin/authorization/rbac/)), are built-in Kubernetes APIs.  APIs are typically used with hosted Kubernetes services, and with managed Kubernetes installations.   They are declarative and use the same conventions are other Kubernetes resources like pods, so new cluster configuration can be repeatable and be managed the same way as applications.  And, where they are stable, they enjoy a [defined support policy](/docs/reference/deprecation-policy/) like other Kubernetes APIs.  For these reasons, they are preferred over **configuration files** and **flags** where suitable.
+
+The remainder of this guide is about Extensions.
+
+## Extensions
+
+Extensions are software components that extend and deeply integrate with kubernetes. 
+They adapt it to support new types and new kinds of hardware.
+
+Most cluster administrators will use a hosted or distribution 
+instance of Kubernetes.  As a result, most Kubernetes users will need to 
+install extensions and fewer will need to author new ones.
+
+## Extension Patterns
+
+Kubernetes is designed to be automated by writing client programs.  Any
+program that reads and/or writes to the Kubernetes API can provide useful
+automation.  **Automation** can run on the cluster or off it.  By following
+the guidance in this doc you can write highly available and robust automation.
+Automation generally works with any Kubernetes cluster, including hosted
+clusters, and managed installations.
+
+There is a specific pattern for writing client programs that work well with
+Kubernetes called the **Controller** pattern.  Controllers typically read an
+object's `.spec`, possibly do things, and then update the object's `.status`.
+
+A controller is a client of Kubernetes.  When Kubernetes is the client and
+calls out to a remote service, it is called a **Webhook.  **The remote service
+is called a **Webhook Backend**.  Like Controllers, Webhooks do add a point of
+failure.  
+
+In the webhook model, Kubernetes makes a network request to a remote service.
+In the **Binary Plugin** model, kubernetes executes a binary (program).
+Binary plugins are used by the kubelet (e.g. [Flex Volume
+Plugins](https://github.com/kubernetes/community/blob/master/contributors/devel/flexvolume.md)
+and [Network
+Plugins](/docs/concepts/cluster-administration/network-plugins/))
+and by kubectl.  Plugin interfaces are still evolving, and may change.
```

Comment:
We shouldn't make any forward looking statements such as "Plugin interfaces are still evolving, and may change."